### PR TITLE
fix(kuma-cp): remove Dataplane for Pod without IP

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
Our pod -> dp reconciler was ignoring reconciliation in two cases

### Pod IP is empty

Our assumption was that if the pod is empty = pod is being created.
The problem is that the Pod can lose its IP, for example - when it's evicted.
In this scenario, we wouldn't reconcile the pod to mark it as unhealthy.

Currently, evicted pods are sometimes marked as unhealthy - when at least one container is down and IP is still there.

Our Dataplane model requires the IP address `Dataplane#networking.address`.
When reconciling pod with empty IP I considered
1) putting some non-routable IP address when Pod IP is missing and always marking this Dataplane as unhealthy
2) retrieving previous networking address and always marking this Dataplane as unhealthy
3) removing Dataplane

I picked option 3. because other options sound like a hack even though technically we won't use unhealthy Dataplane.

### Pod is completed (all containers aside from kuma-sidecar are terminated)

This was added so we remove Dataplane for a finished job in pod status reconciler and not recreate them in pod reconciler.
I was tempted to remove this functionality, but I don't want to introduce a breaking change while fixing other issue.
Instead, I changed the logic to remove Dataplane when Pod Succeeded.
```
// PodSucceeded means that all containers in the pod have voluntarily terminated
// with a container exit code of 0, and the system is not going to restart any of these containers.
PodSucceeded PodPhase = "Succeeded"
```

I also added some additional logging around the pod reconciler.

### Checklist prior to review

- [X] Link to docs PR or issue --
- [X] Link to UI issue or PR --
- [X] Is the [issue worked on linked][1]? --
- [X] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [X] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Unit Tests --
- [X] E2E Tests --
- [X] Manual Universal Tests --
- [X] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch)?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lobkovilya** [comment]
It's kind of nice that today Pod and DPP are always 1 to 1. We don't show Pods in Kuma GUI, so if there is no DPP users may think there is no Pod either (but in reality it's evicted). Can we change the DPP model to allow an empty address only if `health.ready: false`?

**Comment by jakubdyszkiewicz** [comment]
> Can we change the DPP model to allow an empty address only if health.ready: false?

I looked around at the code. This is an interesting idea, but it would be hard to pull this off. There is a lot of code that has the assumption that we have an address. I feel that going in this direction we will generate many bugs.

File: pkg/plugins/runtime/k8s/controllers/pod_controller.go [hunk 1]
```diff
@@ -250,16 +311,18 @@ func (r *PodReconciler) createOrUpdateIngress(ctx context.Context, pod *kube_cor
...
+	log := r.Log.WithValues("pod", kube_types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
 	if err != nil {
-		log := r.Log.WithValues("pod", kube_types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
 		log.Error(err, "unable to create/update Ingress", "operationResult", operationResult)
 		r.EventRecorder.Eventf(pod, kube_core.EventTypeWarning, FailedToGenerateKumaDataplaneReason, "Failed to generate Kuma Ingress: %s", err.Error())
 		return err
 	}
 	switch operationResult {
 	case kube_controllerutil.OperationResultCreated:
+		log.Info("ZoneIngress created")
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/pod_controller.go:322]: Do you think so? I find those operations to be very significant to the system therefore it's nice to have them in logs on INFO level to see a full picture in logs of what happened step by step.

**Comment by jakubdyszkiewicz** [comment]
@mergifyio backport release-1.8

**Pull request merged**